### PR TITLE
fix #20545 assign closure proc to const

### LINKED ISSUE #16736: `const` proc in `let` array causes invalid codegen
### Code
```nim
type ProcType = proc()
const something: ProcType = proc() = discard
let arr: array[1, ProcType] = [something]

```

### Current Output

```
/home/anuke/.cache/nim/bug_d/@mbug.nim.c: In function ‘NimMainModule’:
/home/anuke/.cache/nim/bug_d/@mbug.nim.c:413:99: error: request for member ‘ClE_0’ in something not a structure or union
  413 | lzADdspe4KtMwjNjbL8g[0].ClE_0), colonanonymous___DxAxlW4RjxKP1O7MGX2PnQ.ClE_0);
      |                                                                        ^

/home/anuke/.cache/nim/bug_d/@mbug.nim.c:414:80: error: request for member ‘ClP_0’ in something not a structure or union
  414 | tMwjNjbL8g[0].ClP_0 = colonanonymous___DxAxlW4RjxKP1O7MGX2PnQ.ClP_0;
      |                                                              ^

Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I'/home/anuke/.choosenim/toolchains/nim-#devel/lib' -I/home/anuke -o /home/anuke/.cache/nim/bug_d/@mbug.nim.c.o /home/anuke/.cache/nim/bug_d/@mbug.nim.c' failed with exit code: 1

```

### Expected Output
```
successful compilation, or at least a nim compilation error
```

### Workarounds

- Declaring both the proc and the array as `const` fixes the problem.

### Additional Information

- Occurs in 1.4.2 as well as latest HEAD
- Occurs with and without `--gc:arc`

```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-01-16
Copyright (c) 2006-2021 by Andreas Rumpf

44ceefa9fefca58ba32e9638f754d7f4ca8a2b82 (current git head)
```

### LINKED ISSUE #16737: `const` procs in `const` array are nil in half of array
### Code
```nim
type ProcType = proc()

const a: ProcType = proc() = discard
const arr: array[8, ProcType] = [a, a, a, a, a, a, a, a]

for i in 0..<arr.len:
  echo "[", i, "]: ", arr[i].isNil
echo "\nwith constant index:\n[7]: ", arr[7].isNil
```

### Current Output

```
[0]: false
[1]: false
[2]: false
[3]: false
[4]: true
[5]: true
[6]: true
[7]: true

with constant index:
[7]: false
```

### Expected Output
```
[0]: false
[1]: false
[2]: false
[3]: false
[4]: false
[5]: false
[6]: false
[7]: false

with constant index:
[7]: false
```

### Workarounds

- Declaring both the proc and the array as `var` fixes the problem in this test case. In my codebase, it leads to invalid codegen that I have been unable to isolate.

### Additional Information

- Occurs in 1.4.2 as well as latest HEAD
- Occurs with and without `--gc:arc`
- Likely related to #16736

```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-01-16
Copyright (c) 2006-2021 by Andreas Rumpf

44ceefa9fefca58ba32e9638f754d7f4ca8a2b82 (current git head)
```

### LINKED ISSUE #20544: Calling a proc inside a tuple inside a literal array, Nim compiler generates invalid C code
### What happened?

This gives me compile-time errors from gcc. Two questions:

  - Is this code invalid?
  - And even if this code is invalid, shouldn't the Nim compiler complain, rather than generating C code that gcc rejects?

```
type
  F = proc()
  T = tuple[c: F]

const
  f: F = proc() = echo "Got into the function."
  myTuple: T = (c: f)
  myLiteralArray: array[1, T] = [myTuple]

myLiteralArray[0].c()

```

### Nim Version

Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-10-10
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 1f2075042b7feb06cffe6c4ae6562a2e38144f81
active boot switches: -d:release

### Current Standard Output Logs

_No response_

### Expected Standard Output Logs

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

### LINKED ISSUE #20545: Calling a const proc produces an error message
### What happened?

```
type
  F = proc()

const
  f: F = proc() = echo "Got into f."

proc g() = echo "Got into g."

# Produces this compile-time error message:
#     Error: VM problem: dest register is not set
f()

# Just for contrast: if you comment out the line above, the code compiles
# and this next line runs just fine.
g()
```

### Nim Version

Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-10-10
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 1f2075042b7feb06cffe6c4ae6562a2e38144f81
active boot switches: -d:release

### Current Standard Output Logs

_No response_

### Expected Standard Output Logs

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by bung87**
fix https://github.com/nim-lang/Nim/issues/20545
fix https://github.com/nim-lang/Nim/issues/16736
fix https://github.com/nim-lang/Nim/issues/16737
fix https://github.com/nim-lang/Nim/issues/20544

`proc semConstExpr` fix is by @Araq

### CODE REVIEW

File: tests/closure/tconstclosure.nim [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  cmd: "nim check --hints:off --warnings:off $file"
+  action: "reject"
+  nimout: '''
+tconstclosure.nim(44, 24) Error: const closure proc with an environment cannot be used at runtime: makeMyClosure(myRef)
+
+
+
+
+
+
```
Comment (ringabout) [inline tests/closure/tconstclosure.nim:12]: Why are there lots of blank lines?

Comment (bung87) [reply tests/closure/tconstclosure.nim:12]: I see some test file like this, thought it be usefull when you want add more test cases in same file, and without modify the nimout file line number.

File: compiler/semexprs.nim [hunk 2]
```diff
@@ -873,6 +873,9 @@ proc evalAtCompileTime(c: PContext, n: PNode): PNode =
           localError(c.config, n.info, errCannotInterpretNodeX % renderTree(call))
         else: result = fixupTypeAfterEval(c, result, n)
       else:
+        let noneCtProc = sfCompileTime notin n[0].typ.owner.flags
+        if callee.kind == skConst and (n.typ == nil and noneCtProc):
+          return n
         result = evalConstExpr(c.module, c.idgen, c.graph, call)
```
Comment (Araq) [inline compiler/semexprs.nim:938]: Why can't `evalConstExpr` evaluate procs?

Comment (bung87) [reply compiler/semexprs.nim:938]: it fix `VM problem: dest register is not set` , here it try to eval at compile time, however the const closure is meant to run at runtime. there's no flag distinguish this, so it relys on `n.typ == nil`, yeah, after fix `fixupTypeAfterEval`, I may recheck here?

Comment (bung87) [reply compiler/semexprs.nim:938]: just tried, still can't work without this. maybe should introduce new flag, make this condition simpler.

File: compiler/semexprs.nim [hunk 3]
```diff
@@ -873,6 +873,9 @@ proc evalAtCompileTime(c: PContext, n: PNode): PNode =
           localError(c.config, n.info, errCannotInterpretNodeX % renderTree(call))
         else: result = fixupTypeAfterEval(c, result, n)
       else:
+        let noneCtProc = sfCompileTime notin n[0].typ.owner.flags
+        if callee.kind == skConst and (n.typ == nil and noneCtProc):
```
Comment (Araq) [inline compiler/semexprs.nim]: Why would `n.typ` be nil?

Comment (bung87) [reply compiler/semexprs.nim]: it goes `semConstExpr` -> `fixupTypeAfterEval` -> `semmacrosanity.annotateType` then no branch matches.

**Comment by bung87** [comment]
the disabled test case can be passed by changing to `const G = proc ():int {.compileTime.} =` cc @deech

**Comment by Araq** [comment]
Why add it to the semantic checking phase when we have a dedicated pass for it?

**Comment by bung87** [comment]
which part? or all added lines?

**Comment by Araq** [comment]
All of them.

**Comment by bung87** [comment]
maybe it's closer to the path when I trace the bug, or it can report the error earlier or I didn't find other place do this task.

**Comment by bung87** [comment]
could you point out where to add, I have no idea for now.

**Pull request closed without merging**